File: Problems/leetcode/neetcode roadmap/arrays & hashing/product_of_array_except_self.py
#Vrijeme O(n) bez dijeljenja
nums = [1,2,4,6]

ans = [1]*len(nums)
l, r = 1, 1

# N je umnožak svih predhodnih elemenata
for n in range(len(nums)):
    ans[n] = l
    l *= nums[n]

# Množi N sa umnoškom svih predhodnih elemenata (unazad)
for m in range(len(nums)-1, -1, -1):
    ans[m] *= r
    r *= nums[m]

print(ans)

File: Problems/leetcode/neetcode roadmap/arrays & hashing/product_of_array_except_self.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        int l = nums[0];
        int r = nums[len-1];
        vector<int> ans(len, 1);
        for(int i=1; i<len; i++) {
            ans[i] *= l;
            l *= nums[i];
            ans[len-1-i] *= r;
            r *= nums[len-1-i];
        }
        return ans;
    }
};